src/sage/rings/polynomial/q_integer_valued_polynomials.py

```diff
@@ -0,0 +1,1170 @@
+
+                In the shifted basis, this is the evaluation at `x=0`.
+
+                EXAMPLES::
+
+                    sage: F = QuantumValuedPolynomialRing(ZZ).S()
+                    sage: B = F.basis()
+                    sage: (B[2]*B[4]).sum_of_coefficients()
+                    1
+                """
+                R = self.parent().base_ring()
+                return R.sum(self._monomial_coefficients.values())
+
+    class Shifted(CombinatorialFreeModule, BindableClass):
+        r"""
+        The quantum-valued polynomial ring in the shifted basis.
+
+        The basis used here is given by `S[i] = \genfrac{[}{]}{0pt}{}{i+x}{i}_q` for `i \in \NN`.
+
+        Assuming `n_1 \leq n_2`, the product of two monomials `S[n_1] \cdot S[n_2]`
+        is given by the sum
+
+        .. MATH::
+
+            \sum_{k=0}^{n_1} (-1)^k q^{\binom{k}{2} - n_1 * n_2} \genfrac{[}{]}{0pt}{}{n_1}{k}_q \genfrac{[}{]}{0pt}{}{n_1+n_2-k}{n_1}_q S[n_1 + n_2 - k].
+        """
+        def __init__(self, A):
+            r"""
+            Initialize ``self``.
+
+            EXAMPLES::
+
+                sage: F = QuantumValuedPolynomialRing(QQ).S(); F
+                Quantum-Valued Polynomial Ring over Rational Field
+                in the shifted basis
+                sage: TestSuite(F).run()  # not tested
+            """
+            CombinatorialFreeModule.__init__(self, A.base_ring(),
+                                             NonNegativeIntegers(),
+                                             category=A.Bases(),
+                                             prefix="S",
+                                             latex_prefix=r"\mathbb{S}")
+
+        def _an_element_(self):
+            """
+            Return a small element of ``self``.
+
+            EXAMPLES::
+
+                sage: F = QuantumValuedPolynomialRing(QQ).S()
+                sage: F.an_element()
+                2*S[0] + 4*S[2]
+            """
+            NonNeg = self.basis().keys()
+            ring = self.base_ring()
+            return self.element_class(self, {NonNeg(0): ring(2),
+                                             NonNeg(2): ring(4)})
+
+        def _realization_name(self) -> str:
+            r"""
+            TESTS::
+
+                sage: F = QuantumValuedPolynomialRing(QQ).S()
+                sage: F._realization_name()
+                'shifted'
+            """
+            return "shifted"
+
+        def product_on_basis(self, n1, n2):
+            r"""
+            Return the product of basis elements ``n1`` and ``n2``.
+
+            INPUT:
+
+            - ``n1``, ``n2`` -- integers
+
+            EXAMPLES::
+
+                sage: A = QuantumValuedPolynomialRing(QQ).S()
+                sage: A.product_on_basis(0, 1)
+                S[1]
+            """
+            i = ZZ(n1)
+            j = ZZ(n2)
+            if j < i:
+                j, i = i, j
+            q = self.base_ring().gen()
```

Comment:
Rather than getting this every time, I think it would be better to store `self._q = self.base_ring().gen()` in the `__init__`.

Reply:
The new attribute "_q" is not inherited by the Bases. So I have made a method ".q" there. If you have a better way, please tell me.

Reply:
Either have an ABC for the basis or just have both of them provide it.

Nevertheless, having a method `q()` is a good idea.